grails-app/controllers/org/pih/warehouse/inventory/StockMovementController.groovy

```diff
@@ -283,6 +287,95 @@ class StockMovementController {
+        if (!stockMovement) {
+            flash.message = "${g.message(code: 'default.not.found.message', args: [g.message(code: 'stockMovement.label', default: 'Stock Movement'), params['stockMovement.id']])}"
+            redirect(action: "list")
+            return
+        }
+        Comment comment = Comment.get(params?.id)
+        if (!comment) {
+            flash.message = "${g.message(code: 'default.not.found.message', args: [g.message(code: 'comment.label', default: 'Comment'), comment.id])}"
+            redirect(action: "show", id: stockMovement?.id)
+            return
+        }
+        if (comment.sender.id != session.user.id) {
+            flash.message = "${g.message(code: 'auth.notAuthorized.message')}"
+            redirect(action: "show", id: stockMovement?.id)
+            return
+        }
+        render(view: "addComment", model: [stockMovement: stockMovement, comment: comment])
+    }
+
+    def deleteComment = {
+        def stockMovement = getStockMovement(params['stockMovement.id'])
+        if (!stockMovement) {
+            flash.message = "${g.message(code: 'default.not.found.message', args: [g.message(code: 'stockMovement.label', default: 'Stock Movement'), params['stockMovement.id']])}"
+            redirect(action: "list")
+            return
+        }
+        Comment comment = Comment.get(params?.id)
+        if (!comment) {
+            flash.message = "${g.message(code: 'default.not.found.message', args: [g.message(code: 'comment.label', default: 'Comment'), params.id])}"
+            redirect(action: "show", id: stockMovement?.id)
+            return
+        }
+
+        if (comment.sender.id != session.user.id) {
+            throw new AccessDeniedException("You are not authorized to perform this action")
+        }
+
+        stockMovement.requisition?.removeFromComments(comment)
+        if (!stockMovement.requisition.hasErrors() && stockMovement.requisition?.save()) {
+            flash.message = "${g.message(code: 'default.comments.deleted.label')}"
+            redirect(action: "show", id: stockMovement.id)
+            return
+        }
+        render(view: "show", model: [stockMovement: stockMovement])
+    }
+
+    def saveComment = {
+        def stockMovement = getStockMovement(params['stockMovement.id'])
+        if (!stockMovement) {
+            flash.message = "${g.message(code: 'default.not.found.message', args: [g.message(code: 'stockMovement.label', default: 'StockMovement'), params['stockMovement.id']])}"
+            redirect(action: "list")
+            return
+        }
+        Comment comment = Comment.get(params?.id)
+        if (comment) {
+            if (comment.sender.id != session.user.id) {
+                throw new AccessDeniedException("You are not authorized to perform this action")
+            }
+            comment.properties = params
+            if (!stockMovement.requisition.hasErrors() && stockMovement.requisition.save()) {
```

Comment:
Why you are not saving existing comments directly but you are doing it through requisition (without flush)?

Reply:
I'd say both saves are not necessary if a Comment already exists. Calling save explicilty is only necessary if an entity is not yet persisted, but I don't mind validating it here.
I agree this is weird that it is done through the Requisition. I don't remember if we got rid of disabling the cascade validation (`deepValidation` problem), but if we had not, this would not validate the comment anyway if you did it through the requisitoin.

Reply:
see the comment below - this should be at least moved to the service layer not to experience any problems in Grails 3.

Reply:
For updating a comment, we should be fine with 
```
comment.save()
```
For adding a comment, we should only need to add the comment to the comments association on Requisition and save the requisition.
```
requisition.addToComments(comment)
requisition.save()
```

Reply:
I'd say if we update a comment it should not be even necessary to call the `comment.save()` (but I don't mind doing so), because it's already persisted and hibernate would save it at the end of the session. 
As for second statement, I agree, it would be enough, but if we use the cascade.

Reply:
@kchelstowski the cascade validate problem we encountered was on grails 3. DeepValidate should work on G1